mantidproject/mantid
# Change EnginXFitPeaks behavior when some expected peaks not found

This resolves #11746.

This changes the way errors with found/not found peaks are handled:
- less peaks than expected but enough to fit => notice/warning log but no exception
- only 1 peak: not enough to fit => exception
- no peaks: same, exception

**To test**:
- code review, especially the error messages (see example below to reproduce different issues)
- tests should be fine

This quick example shows the first a third case (you get notices**\* from EnginXFitPeaks for every peak for which the fit fails, and at the end since all have failed you get the exception):

```
sws = Load("ENGINX00228061.nxs")
positions = EnginXCalibrateFull(Workspace=sws, Bank=1, ExpectedPeaks = '1,2,3')
```

**\* could be warnings, but maybe too scary for users, as there's anyway one warning at the end saying that not all the peaks have been fitted, and this can apparently happen for a few peaks but scientists would still check the peaks visually and accept them.

At the same time, since this was dealing with the fitted peaks, I also added an optional output: `OutFittedPeaksTableName` with the fitted peak parameters which will be needed for further work.  We'll need to make this accessible also from the calibration algorithms for the GUI to show the peaks. This example taken from one unit test should produce a table with all the parameters:

```
peak_def1 = "name=BackToBackExponential, I=10000, A=1, B=0.5, X0=8000, S=350"
peak_def2 = "name=BackToBackExponential, I=15000, A=1, B=1.7, X0=15000, S=100"
peak_def3 = "name=BackToBackExponential, I=8000, A=1, B=1.2, X0=20000, S=800"
sws = CreateSampleWorkspace(Function="User Defined", UserDefinedFunction= peak_def1 + ";" + peak_def2 + ";" + peak_def3, NumBanks=1, BankPixelWidth=1, XMin=5000, XMax=30000, BinWidth=25)
EditInstrumentGeometry(Workspace=sws, L2=[1.5], Polar=[90], PrimaryFlightPath=50)
peaksTblName = 'test_fit_peaks_table'
difc, zero = EnginXFitPeaks(sws, WorkspaceIndex=0, ExpectedPeaks=[0.4, 0.83, 1.09], OutFittedPeaksTableName=peaksTblName)
```

And this is a simple example to reproduce the second type of error in the bullet list above:

```
difc, zero = EnginXFitPeaks(sws, WorkspaceIndex=0, ExpectedPeaks=[0.4], OutFittedPeaksTableName=peaksTblName)
```

Error handling is still far from perfect. With bad data, if you fiddle around with the expected peaks you can relatively easily get errors from the Levenberg-Marquardt minimizer (failed to initialize, etc.).
